### ISSUE TITLE
AlertDialog: Allow onPrimaryAction to cancel the closing of the dialog

### ISSUE THREAD

**Issue by tomsontom** [adobe/react-spectrum#1217]
# 🙋 Feature Request

Allow the primaryAction to cancel the closing of the Dialog.

## 🤔 Expected Behavior

Clicking the primary action buttons does not automatically close the dialog

## 😯 Current Behavior

No way of canceling the closing of the dialog

## 💁 Possible Solution

redefine onPrimaryAction to:

```typescript
interface SpectrumAlertDialogProps {
  primaryAction: onPrimaryAction?: () => boolean | void,
}
```

## 🔦 Context

In my application i'd like to have an AlertDialog allowing to execute a remote action and I want to keep the dialog while the remote action is not finished. This is not possible with an AlertDialog today because executing the primary action automatically closes the AlertDialog.

As a workaround I currently use a custom dialog implementation but I'd prefer to use the built-in dialog.

**Comment by LFDanLu** [comment]
@tomsontom I'll bring this up with the team, but I think this behavior is quite intentional as per Spectrum design.

**Comment by LFDanLu** [comment]
Ah, support for loading buttons will be implemented in the future which will cover your use case.

**Comment by tomsontom** [comment]
is there a ticket for that?

**Comment by LFDanLu** [comment]
After further discussion, we'll have to think about this more, the loading button in the ActionDialog might a bit problematic. We'll make a tracking ticket once we discuss later this week and link it here.

**Comment by LFDanLu** [comment]
@tomsontom Hey, so I met with the team and after some discussion we are leaning towards keeping AlertDialog as is since it is meant to be simplistic variant of a standard Dialog with specific behavior. If we were to support the loading state, we would need to add loading props for each of the buttons.  

Could you expand on your use case here (i.e. what remote action are you doing), I could bring it up with the team again. Otherwise going with a standard Dialog would be best here and https://github.com/adobe/react-spectrum/pull/444 will allow you to add a loading state to your button there.

**Comment by tomsontom** [comment]
So my usecase is eg deleting an entry from the database where I wanted to use an AlertDialog of type destructive - as you suggested I created a custom dialog the only disadvantage is that if you change the L&F of the AlertDialog I don't automatically get my dialog aligned but I can see your point of keeping the AlertDialog as simple as possible.

I still think even without a long running action thing there might be reasons one wants to control if the an alert dialog is closed or not but I can't give you a usecase ATM so feel free to close it.

BTW #444 looks exactly what I built myself in the `RemoteActionDialog` so I keep an eye on it!
